File: days/day_07/main.py
import re
import time
import os


def search(bag, rules):
	others = rules[bag]

	if others == {}:
		return False
	else:
		for other in others:
			if other == "shiny gold bag":
				return True
		
		return any(search(other,rules) for other in others)


def count(bag, rules):
	others= rules[bag]

	if others == {}:
		return 0
	else:
		return sum(rules[bag][other]*(1+count(other,rules)) for other in others)


def part_one(rules):
	total = 0

	for bag in rules:
		total += search(bag, rules)
		
	return total


def part_two(rules):
	return count("shiny gold bag", rules)


def main():
	start_time = time.time()

	with open(os.path.dirname(__file__) + '/input.txt', 'r') as data:
		lines = data.readlines()
		rules = {}

		for line in lines:
			bag = re.findall(r'^.*?bag', line)[0]
			rules[bag] = {}

			for item in re.findall(r'[1-9].*?\ bag', line):
				l = item.split(' ')
				num = int(l[0])
				other = ' '.join(l[1:])
				rules[bag][other] = num

		part_one_ans = part_one(rules)
		part_two_ans = part_two(rules)

		print('Day  7 ({:,.3f}s)'.format(time.time()-start_time))
		print('  Part 1: {}'.format(part_one_ans))
		print('  Part 2: {}'.format(part_two_ans))
		

if __name__ == "__main__":
		main()

File: days/day_07/main.js
const fs = require('fs');
const readline = require('readline');

async function loadInput() {
  const rl = readline.createInterface({
    input: fs.createReadStream(__dirname + '/input.txt')
  });

  const lines = [];
  for await (const line of rl) {
    lines.push(line);
  }
  return lines;
}

function search(bag, rules) {
  const others = Object.keys(rules[bag]);

  if (others == []) {
    return false;
  }
  else {
    return (others.some(other => other == "shiny gold bag") 
      || others.some(other => search(other, rules)));
  }

}

function count(bag, rules) {
  const others = Object.keys(rules[bag]);

  if (others.length == 0) {
    return 0;
  }
  else {
    return others.reduce((total, other) => {
     return total + rules[bag][other]*(1+count(other,rules));
    }, 0);
  }
}

function partOne(rules) {
  return Object.keys(rules).reduce((total, bag) => {
    return total + search(bag, rules);
  }, 0);
}

function partTwo(rules) {
  return count("shiny gold bag", rules);
}

async function main() {
	const start_time = process.hrtime();
  const lines = await loadInput();

  const rules = {};
  for (const line of lines){
    const bag = line.match(/^.*?bag/)[0];
    rules[bag] = {};

    for (const item of line.matchAll(/[1-9].*?\ bag/g)) {
      l = item[0].split(' ');
      num = Number(l.shift());
      other = l.join(' ');
      rules[bag][other] = num;
    }
  }

  const part_one_ans = partOne(rules);
  const part_two_ans = partTwo(rules);
  
  const diff = process.hrtime(start_time);
  const total_time = (diff[0] + diff[1]/1e9).toFixed(3);
  
  console.log(`Day  7 (${total_time}s)`);
  console.log(`  Part 1: ${part_one_ans}`);
  console.log(`  Part 2: ${part_two_ans}`);
}

if (require.main === module) {
  main();
}